libs/grok/src/main/java/org/elasticsearch/grok/PatternBank.java

```diff
@@ -0,0 +1,129 @@
+     * In a pattern, anything between <code>%{</code> and <code>}</code> or <code>:</code> is considered
+     * a reference to another named pattern. This method will navigate to all these named patterns and
+     * check for a circular reference.
+     */
+    static void forbidCircularReferences(Map<String, String> bank) {
+        // first ensure that the pattern bank contains no simple circular references (i.e., any pattern
+        // containing an immediate reference to itself) as those can cause the remainder of this algorithm
+        // to recurse infinitely
+        for (Map.Entry<String, String> entry : bank.entrySet()) {
+            if (patternReferencesItself(entry.getValue(), entry.getKey())) {
+                throw new IllegalArgumentException("circular reference in pattern [" + entry.getKey() + "][" + entry.getValue() + "]");
+            }
+        }
+
+        // next, recursively check any other pattern names referenced in each pattern
+        for (Map.Entry<String, String> entry : bank.entrySet()) {
+            String name = entry.getKey();
+            String pattern = entry.getValue();
+            innerForbidCircularReferences(bank, name, new ArrayList<>(), pattern);
+        }
+    }
+
+    private static void innerForbidCircularReferences(Map<String, String> bank, String patternName, List<String> path, String pattern) {
+        if (patternReferencesItself(pattern, patternName)) {
+            String message;
+            if (path.isEmpty()) {
+                message = "circular reference in pattern [" + patternName + "][" + pattern + "]";
+            } else {
+                message = "circular reference in pattern ["
+                    + path.remove(path.size() - 1)
+                    + "]["
+                    + pattern
+                    + "] back to pattern ["
+                    + patternName
+                    + "]";
+                // add rest of the path:
+                if (path.isEmpty() == false) {
+                    message += " via patterns [" + String.join("=>", path) + "]";
+                }
+            }
+            throw new IllegalArgumentException(message);
+        }
+
+        // next check any other pattern names found in the pattern
+        for (int i = pattern.indexOf("%{"); i != -1; i = pattern.indexOf("%{", i + 1)) {
+            int begin = i + 2;
+            int bracketIndex = pattern.indexOf('}', begin);
+            int columnIndex = pattern.indexOf(':', begin);
+            int end;
+            if (bracketIndex != -1 && columnIndex == -1) {
+                end = bracketIndex;
+            } else if (columnIndex != -1 && bracketIndex == -1) {
+                end = columnIndex;
+            } else if (bracketIndex != -1 && columnIndex != -1) {
+                end = Math.min(bracketIndex, columnIndex);
+            } else {
+                throw new IllegalArgumentException("pattern [" + pattern + "] has circular references to other pattern definitions");
```

Comment:
Is this error message correct? Should it say that the pattern has an invalid reference syntax instead of saying it has one that is circular?

Reply:
good catch, even, we're not testing some cases. Added more tests